crates/driver/src/boundary/liquidity/swapr.rs

```diff
@@ -0,0 +1,54 @@
+use {
+    crate::{
+        boundary,
+        domain::liquidity::{self, swapr},
+        infra::{self, blockchain::Ethereum},
+    },
+    anyhow::Result,
+    shared::{
+        current_block::CurrentBlockStream,
+        price_estimation,
+        sources::{swapr::reader::SwaprPoolReader, uniswap_v2::pool_fetching::DefaultPoolReader},
+    },
+    solver::{liquidity::ConstantProductOrder, liquidity_collector::LiquidityCollecting},
+};
+
+/// The base unit for basis points, i.e. how many basis points in 100%.
+const BPS_BASE: u32 = 10_000;
+
+pub fn to_domain(id: liquidity::Id, pool: ConstantProductOrder) -> Option<liquidity::Liquidity> {
+    // invalid Swapr fee ratio; does not have exact BPS representation
+    if (pool.fee.numer() * BPS_BASE) % pool.fee.denom() != 0 {
+        return None;
```

Comment:
I don't like the idea of returning `Option` on error instead of `Result`. `Option` means that a value could be missing, `Result` means that something can go wrong.

Reply:
But I know that you're doing this in some other places as well, ideally the approach should be all-or-nothing. Even if you have to write just a single struct to represent the error I still think that's more correct and expresses your intent more clearly.

Reply:
Your logic is undeniable. Agreed on all fronts.

Reply:
I switched to using unit-structs for representing these errors. I went with an "all-or-nothing" approach.

93024239136c195067de1ceee808179536c27af8
93024239136c195067de1ceee808179536c27af8